### PR TITLE
calls business, connects user and business, user hangs up, user is called back

### PR DESCRIPTION
**Pull request by angel90650**
Goes through the entire process of being connected with business.
uses Twilio conference API
Still needs to be cleaned up
connects our two use cases we showed in the demo, where:
1. the user is connected immediately with the business
2. the user is able to hang up and be called back and be connected with the user.

### CODE REVIEW

File: app/controllers/calls_controller.rb [hunk 1]
```diff
@@ -1,7 +1,14 @@
 class CallsController < ApplicationController
   skip_before_action :verify_authenticity_token
-
-  def begin
+  @@user_number = ""
+  @@has_called = "false"
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: It doesn't look like there's a reason for using a string here. This should be a boolean.

File: app/controllers/calls_controller.rb [hunk 2]
```diff
@@ -10,20 +17,133 @@ def begin
...
+
   private
   def boot_twilio
-  	account_sid = ENV["TWILIO_SID"]
-  	auth_token = ENV["TWILIO_AUTH"]
-  	@client = Twilio::REST::Client.new account_sid, auth_token
+  	#account_sid = ENV["TWILIO_SID"]
+    account_sid = ''
+  	#auth_token = ENV["TWILIO_AUTH"]
+    auth_token = ''
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Why are some lines commented out?

File: app/controllers/calls_controller.rb [hunk 3]
```diff
@@ -10,20 +17,133 @@ def begin
...
+      r.gather(action: '/calls/confirmwait', method: 'POST', numdigits: 2)
+      r.redirect('/calls/rejoinconference')
+    end
+    render xml: response.to_s
+  end
+
+
+
+
+
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Excessive whitespace between methods.

File: app/controllers/calls_controller.rb [hunk 4]
```diff
@@ -10,20 +17,133 @@ def begin
...
+        r.hangup
+    end
+    waitforme
+    render xml: response.to_s
+  end
+  def rejoinconference
+    @@user_callSid = params['CallSid']
+    response = Twilio::TwiML::VoiceResponse.new do |r|
+      r.dial(hangupOnStar: 'true') do |d|
+        d.conference('conference', muted:'False', beep:'False',statusCallbackEvent:'join leave', statusCallback:'/calls/conference',
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Include one space after commas and colons.

File: app/controllers/calls_controller.rb [hunk 5]
```diff
@@ -10,20 +17,133 @@ def begin
...
+  end
+
+  def hangup
+    response = Twilio::TwiML::VoiceResponse.new do |r|
+        r.hangup
+    end
+    waitforme
+    render xml: response.to_s
+  end
+  def rejoinconference
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Include a line between method definitions.

File: app/controllers/calls_controller.rb [hunk 6]
```diff
@@ -10,20 +17,133 @@ def begin
...
+  def announcement
     response = Twilio::TwiML::VoiceResponse.new do |r|
-      r.say(message: 'Forwarding your call now.')
-      r.dial(number: dial_number)
+        r.say(message: 'This is an announcement')
     end
     render xml: response.to_s
   end
 
+
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Extra whitespace.

File: app/controllers/calls_controller.rb [hunk 7]
```diff
@@ -10,20 +17,133 @@ def begin
...
+  def confirmwait
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: `confirmwait` => `confirm_wait`

File: app/controllers/calls_controller.rb [hunk 8]
```diff
@@ -10,20 +17,133 @@ def begin
...
+    input = params['Digits']
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: `input` isn't super descriptive. What does `Digits` represent?

File: app/controllers/calls_controller.rb [hunk 9]
```diff
@@ -10,20 +17,133 @@ def begin
...
+    response = Twilio::TwiML::VoiceResponse.new do |r|
+      if input == '00'
+        r.say(message: 'We will call you back when a business agent is on the line')
+        r.redirect('/calls/hangup', method: 'POST')
+
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Extra new line.

File: app/controllers/calls_controller.rb [hunk 10]
```diff
@@ -10,20 +17,133 @@ def begin
...
-      r.dial(number: dial_number)
+        r.say(message: 'This is an announcement')
     end
     render xml: response.to_s
   end
 
+
+  def confirmwait
+    input = params['Digits']
+    response = Twilio::TwiML::VoiceResponse.new do |r|
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Avoid one letter variable names.

File: app/controllers/calls_controller.rb [hunk 11]
```diff
@@ -10,20 +17,133 @@ def begin
...
+      else
+        r.redirect('/calls/rejoinconference', method: 'POST')
+      end
+    end
+    render xml: response.to_s
+  end
+
+  def hangup
+    response = Twilio::TwiML::VoiceResponse.new do |r|
+        r.hangup
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Inconsistent indentation. Use 2 spaces for indentation.

File: app/controllers/calls_controller.rb (hunk 5 above)
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: `rejoinconference` => `rejoin_conference`

File: app/controllers/calls_controller.rb [hunk 12]
```diff
@@ -10,20 +17,133 @@ def begin
...
+  	#auth_token = ENV["TWILIO_AUTH"]
+    auth_token = ''
+  	@@client = Twilio::REST::Client.new(account_sid, auth_token)
   end
 end
+
+
+#user calls
+#ask for number to call
+#create call from
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: These comments seem incomplete / confusing. Are they leftovers? Can they be removed?

File: app/controllers/calls_controller.rb [hunk 13]
```diff
@@ -10,20 +17,133 @@ def begin
...
+    #call user back
+    call = @@client.calls.create(
+      url: "http://c0ae5a76.ngrok.io/calls/rejoinconference",
+      from: @@dial_number,
+      to: @@user_number
+    )
+    #join conference
+  end
+
+  def announcement
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: This doesn't seem like it'd fit the expected call flow. How is this used?

File: app/controllers/calls_controller.rb [hunk 14]
```diff
@@ -10,20 +17,133 @@ def begin
...
+      if params["CallSid"] == @@user_callSid
+        puts 'user is joining the conference'
+        puts 'there callsid is ' + params['CallSid']
+      end
+      #user = @@client.conferences('conference').participants(@@user_callSid).update(announce_url: "http://c0ae5a76.ngrok.io/calls/announcement")
+    end
+    #render xml: response.to_s
+  end
+
+  def waitforme
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: `waitforme` => `wait_for_me`

File: app/controllers/calls_controller.rb [hunk 15]
```diff
@@ -10,20 +17,133 @@ def begin
...
+    #detect when off hold
+    #call user back
+    call = @@client.calls.create(
+      url: "http://c0ae5a76.ngrok.io/calls/rejoinconference",
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Is this a permanent endpoint?

File: app/controllers/calls_controller.rb [hunk 16]
```diff
@@ -10,20 +17,133 @@ def begin
...
+
+    if @event == "participant-join"
+      puts 'someone is joining the conference'
+      if params["CallSid"] == @@user_callSid
+        puts 'user is joining the conference'
+        puts 'there callsid is ' + params['CallSid']
+      end
+      #user = @@client.conferences('conference').participants(@@user_callSid).update(announce_url: "http://c0ae5a76.ngrok.io/calls/announcement")
+    end
+    #render xml: response.to_s
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Is this necessary? Usually commented out code should be removed.

File: app/controllers/calls_controller.rb [hunk 17]
```diff
@@ -10,20 +17,133 @@ def begin
...
+      r.dial do |d|
+        d.conference('conference', muted:'False', beep:'False',statusCallbackEvent:'join leave', statusCallback:'/calls/conference',
+          statusCallbackMethod:'POST')
+        end
+    end
+    render xml: response.to_s
+  end
+
+  def conference
+    @event = params["StatusCallbackEvent"]
```
Comment (peterwilson) [inline app/controllers/calls_controller.rb]: Does this need to be an instance variable? Usually instance variables in a Rails controller are used for values that need to be shared to the view.

File: app/controllers/calls_controller.rb [hunk 18]
```diff
@@ -1,7 +1,14 @@
 class CallsController < ApplicationController
   skip_before_action :verify_authenticity_token
-
-  def begin
+  @@user_number = ""
+  @@has_called = "false"
+  @@user_callSid = ""
+  @@business_callSid = ""
```
[the remark this answers is not in the document]
Comment (angel90650) [reply app/controllers/calls_controller.rb]: There isn't any kind of state information that is consistent throughout the whole process. Each request only has info that pertains to either the user, business, or conference. 

Would the [low-level caching](https://edgeguides.rubyonrails.org/caching_with_rails.html#low-level-caching) be best to use? 

If we wanted to process multiple sessions, how could we differentiate between sessions in the cache? Every time we get an initial request could we create a unique session id and use that to get the data from cache each time?

**Pull request merged**